File: src/data/courses/dsa/queues/circular-queue/implement-circular-queue/code/java/ImplementCircularQueue.java
class CircularQueue {
    int[] arr;
    int front, rear, size, capacity;
    
    CircularQueue(int k) {
        arr = new int[k];
        capacity = k;
        front = rear = -1;
        size = 0;
    }
    
    boolean enqueue(int x) {
        if (size == capacity) return false;
        if (front == -1) front = 0;
        rear = (rear + 1) % capacity;
        arr[rear] = x;
        size++;
        return true;
    }
    
    boolean dequeue() {
        if (size == 0) return false;
        front = (front + 1) % capacity;
        size--;
        return true;
    }
}

File: src/data/courses/dsa/queues/circular-queue/implement-circular-queue/code/python/implement_circular_queue.py
class CircularQueue:
    def __init__(self, k):
        self.arr = [0] * k
        self.capacity = k
        self.front = self.rear = -1
        self.size = 0
    
    def enqueue(self, x):
        if self.size == self.capacity:
            return False
        if self.front == -1:
            self.front = 0
        self.rear = (self.rear + 1) % self.capacity
        self.arr[self.rear] = x
        self.size += 1
        return True
    
    def dequeue(self):
        if self.size == 0:
            return False
        self.front = (self.front + 1) % self.capacity
        self.size -= 1
        return True